[ci] Run functional tests using bitcoin-qt in one Travis job

jamesob opened the pull request:
As noted in https://github.com/bitcoin/bitcoin/pull/12843, we're currently missing bugs unique to bitcoin-qt during the Travis build because the functional test framework only uses bitcoind (and not bitcoin-qt).

Run a single Travis job under xvfb for framebuffer virtualization and using bitcoin-qt as the binary. I've also removed the `NEED_XVFB` configuration since our usage of xvfb is outdated [per Travis' docs](https://docs.travis-ci.com/user/gui-and-headless-browsers/#Using-xvfb-to-Run-Tests-That-Require-a-GUI) and the corresponding pre-build step is no longer necessary.

jamesob committed:
[ci] Run functional tests using bitcoin-qt in one Travis job

As noted in https://github.com/bitcoin/bitcoin/pull/12843, we're currently
missing bugs unique to bitcoin-qt during the Travis build because the
functional test framework only uses bitcoind (and not bitcoin-qt).

Run a single Travis job under xvfb for framebuffer virtualization and using
bitcoin-qt as the binary. Remove outdated NEED_XVFB Travis configuration.

jnewbery commented:
Concept ACK!

practicalswift commented:
Concept ACK

Sjors commented:
Concept ACK

jamesob commented:
Closing and re-opening in an attempt to get Travis to kick off a build - I was flagged (and then unflagged) as abusing Travis.

fanquake commented on .travis.yml:
```diff
@@ -61,7 +62,6 @@ before_script:
     - if [ -n "$OSX_SDK" -a -f depends/sdk-sources/MacOSX${OSX_SDK}.sdk.tar.gz ]; then tar -C depends/SDKs -xf depends/sdk-sources/MacOSX${OSX_SDK}.sdk.tar.gz; fi
     - if [ -z "$NO_DEPENDS" ]; then make $MAKEJOBS -C depends HOST=$HOST $DEP_OPTS; fi
     # Start xvfb if needed, as documented at https://docs.travis-ci.com/user/gui-and-headless-browsers/#Using-xvfb-to-Run-Tests-That-Require-a-GUI
```
Should remove this comment

jamesob replied:
Thanks! Fixed.

jamesob committed:
Make functional tests compatible with bitcoin-qt usage

In particular, skip any assert_start_raises_init_error checks due to the way
QT handles startup errors (presents the user with an error dialog which must be
closed). In the future we may implement a better way of handling this.

practicalswift commented:
@jamesob I've now restarted the failing Travis build jobs :-)

practicalswift commented:
@jamesob You may want to contact Travis CI support and make sure you're account has not been flagged on their side. That happens and they're usually fixing that quickly after receiving a ticket about it.

laanwj commented:
Concept ACK

conscott commented:
Concept ACK b42a408783b345d3f6c46e26a2d24dd00e1cff70

Seems one test is failing due to #12863. 

For the actual [bitcoin-qt job failing](https://travis-ci.org/bitcoin/bitcoin/jobs/367908554) I tried to reproduce errors locally with Xvfd, manually setting BITCOIND to bitcoin-qt, but had no luck.

jamesob commented:
@conscott thanks for trying to reproduce locally. I've been doing same and while I'm unable to reproduce on my 16.04 host, I've been able to reproduce reliably on a 14.04 VM.

I think running the functional suite using the qt binary has uncovered a latent deadlock, perhaps due to 14.04's older default version of OpenSSL (1.0.1f) based on a tip from @TheBlueMatt. 

Here's some relevant log output from the tests:

```
stderr:
bitcoin-qt: sync.cpp:101: void potential_deadlock_detected(const std::pair<void*, void*>&, const LockStack&, const LockStack&): Assertion `false' failed.
...
node0 2018-04-17T23:41:37.857938Z GUI: PaymentServer::LoadRootCAs: Loaded  145  root certificates 
node0 2018-04-17T23:41:38.262348Z POTENTIAL DEADLOCK DETECTED 
node0 2018-04-17T23:41:38.262376Z Previous lock order was: 
node0 2018-04-17T23:41:38.262386Z  (1) ppmutexOpenSSL[i]  util.cpp:107 
node0 2018-04-17T23:41:38.262403Z  (2) cs  ./addrman.h:508 
node0 2018-04-17T23:41:38.262416Z Current lock order is: 
node0 2018-04-17T23:41:38.262424Z  (2) cs  ./addrman.h:473 
node0 2018-04-17T23:41:38.262437Z  (1) ppmutexOpenSSL[i]  util.cpp:107 
```

I'm going to continue to investigate.

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -83,6 +83,7 @@ def __init__(self):
         self.setup_clean_chain = False
         self.nodes = []
         self.mocktime = 0
+        self.binary = os.getenv("BITCOIND", "bitcoind")
```
Needs rebase due to conflict here

laanwj commented:
this has needed rebase for a long time
closing because of inactivity; let me know if you start work on this again and I'll reopen
